046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,86 @@
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller ends up force restarting the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+ 
+KafkaAgent currently collects the `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is `3`(`RUNNING`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation
+
+To address the issues with restarts and log recovery, KafkaRoller needs to be aware of the current broker state.  If a broker is performing log recovery, KafkaRoller should fail the reconciliation instead of force restarting it.
+
+## Proposal
+ 
+The objective of this proposal is to introduce a REST API endpoint within KafkaAgent, which exposes metrics from a Kafka broker. Furthermore, KafkaRoller would be modified so that it can query this endpoint to determine the current state of a broker, thus preventing unnecessary restarts. Additionally, KafkaRoller would also display the remaining logs and segments to recover, allowing a human operator to observe the recovery process and determine if it is making progress or is stuck.
+
+### Implementation details
+
+*KafkaAgent*:
+- Add a web server to accept SSL connections on port 8443.
+- Use Jetty for the web server since Kafka already uses it therefore avoids an additional dependency.
+- Reuse the broker's own certificate that would be in the container for TLS authentication. 
+- Configure the server to require TLS client authentication. 
+- Expose an endpoint `v1/brokermetrics` that includes the API version in the URI path.
+- Extend KafkaAgent to collect `kafka.log:type=LogManager,name=remainingLogsToRecover` and `kafka.log:type=LogManager,name=remainingSegmentsToRecover` metrics.
+- Take the metric name as a parameter in the request query e.g `v1/brokermetrics?name=BrokerState`. Use the value of the MBean's `name` field for the request  e.g. `BrokerState` and  `remainingLogsToRecover`.
+- Use Jetty's native ServerHandler API to handle a `GET` request and return 
+    - an integer value for `BrokerState`.
+    - an integer value for  `remainingLogsToRecover`.
+    - an integer value for `remainingSegmentsToRecover`.
+- If no metric name was provided in the request or the requested metric name is not supported, return HTTP 404 (Not Found) to the client.
+
+*KafkaRoller*:
+- Use the Pod DNS name to query the KafkaAgent API.
+- Use the cluster operator's client certificate and keys to authenticate with the KafkaAgent API server.
+- Instead of checking the broker's readiness through `isReady()` function, utilise the new RESTful endpoint to determine if the broker is ready.
```

Comment:
Should it be really instead? Or in addition? Or some other combination? (e.g. after we are done with the HTTP endpoint we check the readiness)? The readiness is important so it should still be checked that it eventually gets ready.

Reply:
In future considerations I read that readiness probe will use the agent's endpoint. I think it's better if we make it part of this proposal, while keeping the old behavior for backward compatibility.

Reply:
Now thinking about it, it does not make sense to change the readiness function in KafkaRoller. 

The readiness probe part was put in the future consideration because it's not really a part of solving the log recovery issue. I think once we have the agent exposing the API, it should be in a separate proposal.